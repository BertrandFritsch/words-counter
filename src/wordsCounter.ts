interface Words {
  [ word: string ]: number;
}

const wordRegex = /[A-Za-zÀ-ÿ-]{3,}/g;

/**
 * Counts the number of occurrences of the words in a text.
 * The function recognizes french language characters.
 *
 * The words with less than 2 letters are ignored.
 *
 * Assumes that the text is well formed. The function doesn't try
 * to reject badly formatted words.
 *
 * @param {string} text
 * @returns {Words} A property bag containing the words with their frequency
 */
export function count(text: string): Words {
  return (text.match(wordRegex) || []).reduce((wordCounts, word) => {

    wordCounts[ word ] = (wordCounts[ word ] || 0) + 1;
    return wordCounts;

  }, {} as Words);
}
