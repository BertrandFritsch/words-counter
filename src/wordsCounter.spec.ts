import * as fs from 'fs';
import * as wordsCounter from './wordsCounter';

describe('words counter module', () => {
  it('should get an empty map for an empty string', () => {
    expect(wordsCounter.count('')).toEqual({});
  });

  it('should count 1 for a single word', () => {
    expect(wordsCounter.count('word')).toEqual({ word: 1 });
  });

  it('should count 2 for twice the same word', () => {
    expect(wordsCounter.count('word word')).toEqual({ word: 2 });
  });

  it('should ignore words less or equal to 2 letters', () => {
    expect(wordsCounter.count('le mot le plus long')).toEqual({ mot: 1, plus: 1, long: 1 });
  });

  it('should accept more than 1 space between 2 words', () => {
    expect(wordsCounter.count('word      time')).toEqual({ word: 1, time: 1 });
  });

  it('should accept multi-line text', () => {
    expect(wordsCounter.count('word\ntime')).toEqual({ word: 1, time: 1 });
  });

  it('should ignore punctuation', () => {
    expect(wordsCounter.count('word, time. next; exclamation! interrogation! dot dot dot...')).toEqual({ word: 1, time: 1, next: 1, exclamation: 1, interrogation: 1, dot: 3 });
  });

  it('should recognize french accents', () => {
    expect(wordsCounter.count('être')).toEqual({ être: 1 });
  });

  it('should recognize dash separated words', () => {
    expect(wordsCounter.count('peut-être')).toEqual({ 'peut-être': 1 });
  });

  it('should recognize uppercase letters', () => {
    expect(wordsCounter.count('PEUT-ÊTRE')).toEqual({ 'PEUT-ÊTRE': 1 });
  });

  it('should recognize the single cote as a separator', () => {
    expect(wordsCounter.count('Jean d\'Angleterre')).toEqual({ Jean: 1, Angleterre: 1 });
  });

  it ('should handle a whole text', () => {
    const text = fs.readFileSync('./data/text.txt', 'utf8');
    expect(wordsCounter.count(text)).toMatchSnapshot();
  });
});
