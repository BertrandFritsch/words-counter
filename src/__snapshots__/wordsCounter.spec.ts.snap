// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`words counter module should handle a whole text 1`] = `
Object {
  "Aliénor": 1,
  "Angleterre": 3,
  "Aquitaine": 1,
  "Arthur": 1,
  "Bois": 1,
  "Bouvines": 1,
  "Carta": 1,
  "Cela": 1,
  "Ces": 1,
  "Cinquième": 1,
  "Considéré": 1,
  "Couronne": 1,
  "Empire": 1,
  "Est": 1,
  "France": 3,
  "Geoffroy": 1,
  "Guerre": 1,
  "Guillaume": 1,
  "Henri": 4,
  "III": 2,
  "Ier": 1,
  "Innocent": 1,
  "Irlande": 1,
  "Irrités": 1,
  "Jean": 7,
  "Les": 1,
  "Louis": 1,
  "Magna": 1,
  "Malgré": 1,
  "Normandie": 1,
  "Philippe": 1,
  "Plantagenêt": 1,
  "Première": 1,
  "Richard": 1,
  "Robin": 1,
  "Shakespeare": 1,
  "Terren": 1,
  "Tudor": 1,
  "VIII": 1,
  "accession": 1,
  "accorda": 1,
  "accroissant": 1,
  "actuel": 1,
  "administrateur": 1,
  "affligé": 1,
  "affronter": 1,
  "ainsi": 1,
  "alliances": 1,
  "alliés": 1,
  "alors": 2,
  "angevins": 1,
  "anglais": 1,
  "année": 1,
  "apaisa": 1,
  "appliqué": 1,
  "après": 2,
  "armée": 1,
  "ascendant": 1,
  "aspects": 1,
  "avec": 1,
  "aventures": 1,
  "aînés": 1,
  "barons": 2,
  "base": 1,
  "bloquée": 1,
  "campagne": 1,
  "ces": 1,
  "changea": 1,
  "comme": 2,
  "comportement": 1,
  "compétent": 1,
  "compétition": 1,
  "confronté": 1,
  "consacra": 1,
  "consensus": 1,
  "considérablement": 1,
  "contemporains": 1,
  "continent": 1,
  "continentale": 1,
  "continentaux": 1,
  "contraint": 1,
  "contre": 1,
  "croisade": 1,
  "cruelle": 1,
  "culture": 1,
  "dans": 1,
  "depuis": 1,
  "dernier": 1,
  "des": 8,
  "destiné": 1,
  "destinés": 1,
  "devint": 2,
  "dispositions": 1,
  "dispute": 1,
  "donc": 1,
  "droits": 1,
  "dut": 1,
  "dysenterie": 1,
  "débats": 1,
  "défaite": 1,
  "définitivement": 1,
  "efforts": 1,
  "entraîna": 1,
  "entre": 1,
  "est": 1,
  "excommunication": 1,
  "faisait": 1,
  "fait": 1,
  "fiction": 1,
  "fils": 3,
  "finalement": 1,
  "financer": 1,
  "fit": 2,
  "formant": 2,
  "forte": 1,
  "frère": 1,
  "frères": 2,
  "fut": 5,
  "garantissant": 1,
  "grande": 1,
  "général": 1,
  "hausse": 1,
  "historiens": 2,
  "historiques": 1,
  "hommes": 1,
  "héritage": 1,
  "héritier": 1,
  "héros": 1,
  "immédiatement": 1,
  "impôts": 1,
  "influent": 1,
  "jugé": 1,
  "les": 11,
  "libres": 1,
  "lui": 2,
  "mais": 2,
  "manque": 1,
  "menace": 1,
  "militaires": 1,
  "monarque": 1,
  "monter": 1,
  "mort": 2,
  "mourut": 1,
  "méprisant": 1,
  "méprisante": 1,
  "neveu": 1,
  "nobles": 2,
  "nombreuses": 1,
  "nombreux": 1,
  "notamment": 2,
  "nouveau": 1,
  "nouvelle": 1,
  "négatifs": 1,
  "objet": 1,
  "offensive": 1,
  "ont": 3,
  "opposition": 1,
  "pape": 1,
  "par": 8,
  "participait": 1,
  "partie": 1,
  "pas": 1,
  "perdit": 1,
  "permit": 1,
  "personnage": 1,
  "personnalité": 1,
  "peu": 1,
  "plus": 1,
  "poitevins": 1,
  "politique": 1,
  "populaire": 1,
  "posée": 1,
  "pouvoir": 1,
  "prendre": 2,
  "prince": 1,
  "proto-protestant": 1,
  "préféré": 1,
  "présenté": 1,
  "père": 1,
  "que": 1,
  "quelconque": 1,
  "qui": 3,
  "raison": 2,
  "rapidement": 1,
  "ratée": 1,
  "rebelles": 2,
  "recevoir": 1,
  "reconquérir": 1,
  "rentrer": 1,
  "respectèrent": 1,
  "ressources": 1,
  "reste": 1,
  "retour": 1,
  "revenus": 1,
  "roi": 4,
  "royaume": 1,
  "règne": 2,
  "réformant": 1,
  "révolte": 1,
  "révoltèrent": 1,
  "sans": 2,
  "seigneur": 1,
  "selon": 1,
  "servi": 1,
  "ses": 7,
  "signature": 1,
  "situation": 1,
  "solda": 1,
  "son": 8,
  "soutenus": 1,
  "souverain": 1,
  "successeur": 1,
  "succès": 1,
  "suivante": 1,
  "sur": 4,
  "surnommé": 1,
  "taxes": 1,
  "tensions": 1,
  "tenta": 1,
  "tenter": 1,
  "terres": 1,
  "territoire": 1,
  "territoires": 2,
  "traitement": 1,
  "trois": 1,
  "troisième": 1,
  "trône": 2,
  "tyran": 1,
  "tyrannique": 1,
  "une": 2,
  "uvres": 1,
  "valut": 1,
  "varié": 1,
  "via": 1,
  "victorienne": 1,
  "éclata": 1,
  "également": 1,
  "époque": 1,
  "époques": 1,
  "était": 1,
  "été": 1,
  "évaluations": 1,
}
`;
